### ISSUE TITLE
Use of SmartPointers in container-like types like MSExperiment, ConsensusMap, ...

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#4800]
Wrapping of MSExperiment, ConsensusMap, etc. in pyOpenMS leads to very inefficient access.
E.g. calling getSpectra() will copy all spectra as individual python objects.
The question would be how to properly replace those by smart pointers (and what kind of smart pointers)
and discuss the pros and cons.

**Comment by timosachsenberg** [comment]
My guess is that we would use a reference-counted smart pointer (as in boost or now in std).
so how would those methods look like:
```
    const std::vector<MSSpectrum>& getSpectra() const;
    std::vector<MSSpectrum>& getSpectra();
    MSSpectrum& getSpectrum(Size id);
```
would those return/expect an std::smart_ptr<MSSpectrum> or container thereof?

**Comment by cbielow** [comment]
advantage of changing the interface:
1) external python bindings get faster (by how much?)

disadvantages
1) our C++ interface does surprising things (there is not reason to pass around smart pointers other than Python bindings). One should not surprise people! In other words: the interface is non-standard and people will waste brain-resources to figure out why. Not good.
2) The code will be slower. By how much needs to be seen.
3) Compiling will be slower (smart pointers are real beasts)

So, in summary, not a good idea.
Can Python handle a plain pointer? The current notion is: if you get a raw pointer, memory management is not your business, otherwise they would have given you a smart pointer.
If really need be, one could write a Wrapper class, which internally holds a reference to an MSExperiment and has smart-pointer accessors. This way, MSExperiment would not change, and Python-people can just do
`SmartMSExperiment(exp).getSpectra()` and get a smart pointer.

**Comment by timosachsenberg** [comment]
SmartMSExperiment(exp).getSpectra() looks a bit awkward from a user perspective.
Considering that we don't have too many classes and methods where this is relevant could we also use the smart pointer versions if pyopenms is compiled?

**Comment by jpfeuffer** [comment]
Maybe you can wrap SmartMsExp as MsExp. Would be preferable over Cmake macros, making the default configured library incompatible with pyOpenMS.

**Comment by hendrikweisser** [comment]
Can I just state for the record that I'm vehemently against making the C++ library "worse" (more complex, less efficient...) to deal with limitations of the Python bindings?

So I agree with Chris that using a wrapper class, or in the worst case extending the C++ API with additional "Python-compliant" functions, would be a better way forward.

**Comment by hroest** [comment]
I would like to add that there are also benefits in the C++ code for using pointers, for example it would be much faster to sort an MSExperiment if all you have to do is sort a list of pointers instead of sorting a list of heavy objects. 

The main advantage for the Python bindings is not really that they get faster, but that they are currently quite suprising:

```
s = exp.getSpectrum(5)
s.setRT(5.0)
```

does something surprising: it changes a copy of the MSSpectrum object number 5 and the underlying spectrum in the `exp` is unchanged. This I think is the real problem, currently pyOpenMS is in a sense "read only" since you have to remember a lot of stuff to make it work correctly, eg put the spectrum back at position 5 after changing it.

@cbielow yes Python can handle a plain pointer but the trouble is ownership. Since it is not clear who "owns" a raw pointer, pyOpenMS will then copy the underlying data into a smart pointer that it now owns and knows cannot go away under its nose. Since C++ does ownership implicitly (documentation) and not explicitly (smart_ptr) we cannot really change that safely. I think the wrapper functions are the *worst* possible solutions, because what happens when the smart_ptr reference count reaches zero? It will attempt to delete the underlying object, so whenever an object that got passed to Python goes out of scope, the underlying C++ object may get deleted. 

My points would be 

1. Currently the Python behavior *is* surprising
2. Using (smart) pointers is *not* necessarily surprising to a C++ programmer (should not be)
3. I do not really believe that the code will noticeably slower if we change MSExperiment to a `vector<smart_ptr<MSSpectrum>>` due to the current overhead of MSSpectrum

**Comment by cbielow** [comment]
> I would like to add that there are also benefits in the C++ code for using pointers, for example it would be much faster to sort an MSExperiment if all you have to do is sort a list of pointers instead of sorting a list of heavy objects.

Unfortunately, probably not true, since copying smart-ptrs costs you an atomic operation, which is far from free (write cache flushes and other ugly things). Sorting uses move semantics, which (depending on the items being moved) is fast! Also, since its a pointer, there is yet another layer of indirection, which is bad for cache performance. All in all, on the spectrum level, it probably does not harm much, since the data progessing itself takes way longer, but speed is definitely not something which smart pointers are known for! (in fact, sorting might be their only stronghold, but it's nowhere near critical).

> 
> The main advantage for the Python bindings is not really that they get faster, but that they are currently quite suprising:
> 
> ```
> s = exp.getSpectrum(5)
> s.setRT(5.0)
> ```
> 
> does something surprising: it changes a copy of the MSSpectrum object number 5 and the underlying spectrum in the `exp` is unchanged. 

Being more intricate with R, this is exactly what I would expect, since everything is a copy. But I see that this is ugly, requires more boilerplate code and is error prone.

This I think is the real problem, currently pyOpenMS is in a sense "read only" since you have to remember a lot of stuff to make it work correctly, eg put the spectrum back at position 5 after changing it.
> 
> @cbielow yes Python can handle a plain pointer but the trouble is ownership. Since it is not clear who "owns" a raw pointer, 

The general notion is that, if someone passes you a raw pointer, you do not own it. The object you got it from does. Now, you have to take care to keep the parent object around for at least as long as the raw pointer, but usually, you pass raw pointers into functions, and the enclosing scope outlives the function scope, so everything is fine (unless you do stupid things and store the pointer somewhere else)...

> pyOpenMS will then copy the underlying data into a smart pointer that it now owns and knows cannot go away under its nose. Since C++ does ownership implicitly (documentation) and not explicitly (smart_ptr) we cannot really change that safely. I think the wrapper functions are the _worst_ possible solutions, because what happens when the smart_ptr reference count reaches zero? It will attempt to delete the underlying object, so whenever an object that got passed to Python goes out of scope, the underlying C++ object may get deleted.
> 

I probably phrased that wrongly, but in essence: make a copy once, and hand out smart pointers as you want, but do it in python.

> My points would be
> 
> 1. Currently the Python behavior _is_ surprising

That may be the case, but we should not cripple a C++ library, just to make python bindings happy. There must be another way.

> 2. Using (smart) pointers is _not_ necessarily surprising to a C++ programmer (should not be)

Oh in this case it definitely is. There is simply no reason for it.

> 3. I do not really believe that the code will noticeably slower if we change MSExperiment to a `vector<smart_ptr<MSSpectrum>>` due to the current overhead of MSSpectrum

Maybe, maybe not. But you are willing to cripple on interface in the native library to make a second class citizen (python) less surprising. Imagine range-based for-loops over spectra. What you get are smart pointers, which you iterate over. This is a) not common whatsoever, b) surprising c) confusing, because `for (const auto& smart : exp.getSpectra()) does not mean what it should. And how do you now write a for-loop over `const MSSpectrum`? See the problems? Furthermore, changing the interface will cause a lot of code to break -- the time spend fixing this would be much wiser spend on other things.

Now, let me say again: find a solution in cython/pyOpenMS, either by documentation (use raw pointers) or some other wrapping mechanism (I am sure you can think of one).

**Comment by timosachsenberg** [comment]
What is your opinion on solving the issue on the C++ side by having two different SpectrumTypes depending on whether we compile OpenMS or pyOpenMS? I think this has not been mentioned yet.

**Comment by jpfeuffer** [comment]
This will probably timeout bioconda because we have to compile the lib two times, keep that in mind. You will have to split conda builds again.

**Comment by hroest** [comment]
> What is your opinion on solving the issue on the C++ side by having two different SpectrumTypes depending on whether we compile OpenMS or pyOpenMS? I think this has not been mentioned yet.

@timosachsenberg  still seems tricky since the rest of the library will still work, its not just that you can replace the member `vector<MSSpectrum>` with `vector<smart_ptr<MSSpectrum>>` and everything will work -- since everything will break. This is so central in OpenMS that it will be hard to keep both versions working. 

@cbielow I am trying to think about your solution but lets assume we have two Python functions like this

```
def myFxn1(filename):
  tmp = MSExperiment()
  mzMLFile().load(filename, tmp)
  tmp.getSpectrum(0).setRT(5)
  s = tmp.getSpectrum(0)
  return s

def myFxn2(filename):
  tmp = MSExperiment()
  for k in range(n): 
    s = MSSpectrum()
    tmp.push_back(s)
  return tmp
```

If you are a C++ programmer and you see `myFxn1`, then you look up `getSpectrum` to see what it returns (a copy, a ref, const-ref, ptr or smart_ptr). If you are Python programmer you dont do that, you assume it "just works" (basically everything in Python is a smart_ptr, it uses garbage collection internally). I think the problem is that we are trying to find a solution for a language (Python) where people simply never think about memory management and we have to make the interface such that people find it "native". 

> The general notion is that, if someone passes you a raw pointer, you do not own it. The object you got it from does. Now, you have to take care to keep the parent object around for at least as long as the raw pointer, but usually, you pass raw pointers into functions, and the enclosing scope outlives the function scope, so everything is fine (unless you do stupid things and store the pointer somewhere else)...

Lets say there is now a fxn `MSSpectrum * getSpectrum(int)` which gets used here, returns a raw ptr. How do we code that in Python? The `Cython_MSSpectrum` object will now hold a raw ptr to a C++ object of type `MSSpectrum` and needs ownership information (in this case it doesnt own it, so it should not be destroyed once its out of scope). Secondly, how will the `Cython_MSSpectrum` object know if its safe to access the underlying C++ object since the data could have gone out of scope and could be destroyed (as in the example above). On the other hand it also needs to deal with the case that it *does* own it, eg if you create a new  `Cython_MSSpectrum` *in Python* then the underlying C++ object should be be destroyed if the Python object goes out of scope.

> Being more intricate with R, this is exactly what I would expect, since everything is a copy. But I see that this is ugly, requires more boilerplate code and is error prone.

this is not expected in Python since everything is a smart_ptr basically

> I probably phrased that wrongly, but in essence: make a copy once, and hand out smart pointers as you want, but do it in python.

I dont think that would work either, since we need the original MSExperiment to stick it back into OpenMS, otherwise this would never work:

```
  tmp = MSExperiment()
  mzMLFile().load(filename, tmp)
  tmp.getSpectrum(0).setRT(5)
  FeatureFinderCentroided().run( tmp )
```

where the last line needs a C++ object of type `MSExperiment` to work properly. (Note line 3 is a NOP right now). So I think the wrapper class idea does not really solve our problem either. (otherwise all fxns that accept `MSExperiment` need also to be patched to also accept the wrapper class).

**Comment by hroest** [comment]
from todays discussion, maybe we can use a "view" approach in Python:

```
tmp.getSpectrum(0).setRT(5) 
tmp.getSpectrum(0).getPrecursor().setMZ() 
tmp.getSpectrumView(0).getPrecursorView().setMZ(5.0) 
```

**Comment by jpfeuffer** [comment]
I thought about this recently and it should be relatively easy to support views in autowrap by giving the wrapped classes a parent attribute that keeps the original owner of the data alive.
However this does not save you from e.g getting a reference to a Spectrum from an Experiment and then changing the Experiment by e.g. adding a few spectra such that it reallocates the vector and the Spectrum reference then points to deallocated space. We would need to forbid certain methods while a parent class has views on it. Or only allow views for parents of constant compositions..

**Comment by timosachsenberg** [comment]
this is largely solved with the nanobind bindings

**Comment by jpfeuffer** [comment]
@timosachsenberg Is it? Maybe but I don't think nanobind handles any of this by default. You need to tell it to wrap things as refs or copies and I don't think it does ref counting or backref to parents by default.
I think you will have similar problems as before.

**Comment by timosachsenberg** [comment]
yeah but we probably don't want to use smart ptr
